firedrake/slate/slac/kernel_builder.py

```diff
@@ -1,76 +1,70 @@
+from abc import ABCMeta, abstractmethod, abstractproperty
 
 from coffee import base as ast
 
-from firedrake.slate.slate import (TensorBase, Tensor, TensorOp,
-                                   Action, Inverse)
-from firedrake.slate.slac.utils import (Transformer, traverse_dags,
-                                        collect_reference_count)
+from collections import OrderedDict
+
+from firedrake.slate.slate import (TensorBase, Tensor,
+                                   TensorOp, Action,
+                                   Negative, Transpose)
+from firedrake.slate.slac.utils import (traverse_dags,
+                                        collect_reference_count,
+                                        count_operands)
 from firedrake.utils import cached_property
 
 from ufl import MixedElement
 
 
-class KernelBuilder(object):
-    """A helper class for constructing Slate kernels.
-
-    This class provides access to all temporaries and subkernels associated
-    with a Slate expression. If the Slate expression contains nodes that
-    require operations on already assembled data (such as the action of a
-    slate tensor on a `ufl.Coefficient`), this class provides access to the
-    expression which needs special handling.
+class KernelBuilderBase(object, metaclass=ABCMeta):
+    """A base helper class for constructing Slate kernels."""
 
-    Instructions for assembling the full kernel AST of a Slate expression is
-    provided by the method `construct_ast`.
-    """
     def __init__(self, expression, tsfc_parameters=None):
-        """Constructor for the KernelBuilder class.
+        """Constructor for the KernelBuilderBase class.
 
         :arg expression: a :class:`TensorBase` object.
         :arg tsfc_parameters: an optional `dict` of parameters to provide to
                               TSFC when constructing subkernels associated
                               with the expression.
         """
         assert isinstance(expression, TensorBase)
+
+        # Collect reference counts
+        ref_counts = collect_reference_count([expression])
+
+        # Collect terminals and expressions
+        temps = OrderedDict()
+        action_coefficients = []
+        tensor_ops = []
+        for tensor in traverse_dags([expression]):
+            if isinstance(tensor, Tensor):
+                temps.setdefault(tensor, ast.Symbol("T%d" % len(temps)))
+
+            elif isinstance(tensor, TensorOp):
+                # Actions will always require a coefficient temporary.
+                if isinstance(tensor, Action):
+                    actee, = tensor.actee
+                    action_coefficients.append(actee)
+
+                # Operations which have "high" reference count will have
+                # auxiliary temporaries created. Negative and Transpose
+                # operations will not have extra temporaries.
+                if ref_counts[tensor] > 1 and not isinstance(tensor, (Negative, Transpose)):
```

Comment:
Is it obvious why you don't make a temporary for a transposed matrix?

Reply:
Well, in Eigen `A.transpose()` returns a proxy object without actually computing the transpose. Assignments will trigger an evaluation: `T = A.transpose()`. However, now that I think about it, if a transpose appears in n places, (a transpose proxy object appears n times in the expression), then surely the transpose would need to physically be evaluated at some point? So this means n transpose evaluations?

OK. I think I'll go ahead and amend this and include the transpose.